src/peergos/server/space/JdbcUsageStore.java

```diff
@@ -0,0 +1,370 @@
+            throw new RuntimeException(sqe);
+        }
+    }
+
+    private PublicKeyHash getWriter(int writerId) {
+        try (PreparedStatement search = conn.prepareStatement("SELECT key_hash FROM writers WHERE id = ?;")) {
+            search.setInt(1, writerId);
+            ResultSet resultSet = search.executeQuery();
+            resultSet.next();
+            return PublicKeyHash.decode(resultSet.getBytes(1));
+        } catch (SQLException sqe) {
+            LOG.log(Level.WARNING, sqe.getMessage(), sqe);
+            throw new RuntimeException(sqe);
+        }
+    }
+
+    @Override
+    public WriterUsage getUsage(PublicKeyHash writer) {
+        String owner = getOwner(writer);
+        int writerId = getWriterId(writer);
+        Set<PublicKeyHash> owned = new HashSet<>();
+        try (PreparedStatement search = conn.prepareStatement("SELECT owned_id FROM ownedkeys WHERE parent_id = ?;")) {
+            search.setInt(1, writerId);
+            ResultSet resultSet = search.executeQuery();
+            while (resultSet.next())
+                owned.add(getWriter(resultSet.getInt(1)));
+        } catch (SQLException sqe) {
+            LOG.log(Level.WARNING, sqe.getMessage(), sqe);
+            throw new RuntimeException(sqe);
+        }
+        try (PreparedStatement search = conn.prepareStatement("SELECT target, direct_size FROM writerusage WHERE writer_id = ?;")) {
+            search.setInt(1, writerId);
+            ResultSet resultSet = search.executeQuery();
+            resultSet.next();
+            MaybeMultihash target = Optional.ofNullable(resultSet.getBytes(1))
+                    .map(x -> MaybeMultihash.of(Cid.cast(x)))
+                    .orElse(MaybeMultihash.empty());
+            return new WriterUsage(owner, target, resultSet.getLong(2), owned);
+        } catch (SQLException sqe) {
+            LOG.log(Level.WARNING, sqe.getMessage(), sqe);
+            throw new RuntimeException(sqe);
+        }
+
+    }
+
+    @Override
+    public void updateWriterUsage(PublicKeyHash writer,
+                                  MaybeMultihash target,
+                                  Set<PublicKeyHash> ownedKeys,
+                                  long retainedStorage) {
+        String owner = getOwner(writer);
+        int writerId = getWriterId(writer);
+        try (PreparedStatement insert = conn.prepareStatement("UPDATE writerusage SET target=?, direct_size=? WHERE writer_id = ?;")) {
+            insert.setBytes(1, target.isPresent() ? target.get().toBytes() : null);
+            insert.setLong(2, retainedStorage);
+            insert.setInt(3, writerId);
+            int count = insert.executeUpdate();
+            if (count != 1)
+                throw new IllegalStateException("Didn't update one record!");
+        } catch (SQLException sqe) {
+            LOG.log(Level.WARNING, sqe.getMessage(), sqe);
+            throw new RuntimeException(sqe);
+        }
+        setWriters(owner, writer, ownedKeys);
```

Comment:
same deal as the other place. I guess if the call to updateWroterUsage() for the same writerId is never concurrent, then it should be ok.